HydraConfig support

otherman16 opened the pull request:
Hydra support
1. MNIST classification test pipeline was added
2. catalyst/experiments/hydra_config.py: HydraConfigExperiment was added
3. catalyst/runners/hydra_supervised.py: HydraSupervisedExperiment was added
4. catalyst/dl/script/hydra_run.py: @hydra.main() was added as entry point for hydra configuration experiment
5. catalyst/dl/script/run.py: optional argument '--hydra' was added to change entrypoint to hydra config experiment one
6. catalyst/settings.py: optional 'hydra' package import was added
7. catalyst/utils/hydra_config.py: hydra config utils was added
8. catalyst/utils/hydra_scripts.py: hydra scripts utils was added
9. catalyst/utils/hydra_sys.py: hydra sys utils was added
10. requirements/requirements-hydra.txt: optional 'hydra-core' package was added

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by blank line to reduce collisions -->


## Description

<!-- Add a more detailed description of the changes if needed. -->
Hydra support for catalyst-dl run

Previous PR: [OmegaConf support for "catalyst-dl run" #943](https://github.com/catalyst-team/catalyst/pull/943)


## Related Issue

<!-- If your PR refers to a related issue, link it here. -->
[Hydra/OmegaConf configuration structure #912](https://github.com/catalyst-team/catalyst/issues/912)


## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)


## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

<!-- Thank you for your contribution! -->


PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

otherman16 committed:
1. tests/_tests_cv_classification_hydra: 'catalyst-dl run --hydra' MNIST classification test pipeline was added
2. catalyst/experiments/hydra_config.py: HydraConfigExperiment was added
3. catalyst/runners/hydra_supervised.py: HydraSupervisedExperiment was added
4. catalyst/dl/script/hydra_run.py: @hydra.main() was added as entry point for hydra configuration experiment
5. catalyst/dl/script/run.py: optional argument '--hydra' was added to change entrypoint to hydra config experiment one
6. catalyst/settings.py: optional 'hydra' package import was added
7. catalyst/utils/hydra_config.py: hydra config utils was added
8. catalyst/utils/hydra_scripts.py: hydra scripts utils was added
9. catalyst/utils/hydra_sys.py: hydra sys utils was added
10. requirements/requirements-hydra.txt: optional 'hydra-core' package was added

otherman16 committed:
Update hydra_supervised.py, __init__.py, and transforms.py

otherman16 committed:
Update hydra_run.py, run.py, and 3 more files...

Scitator commented on bin/tests/check_dl_cv.sh:
```diff
@@ -603,5 +603,61 @@ fi
+LOGDIR=./tests/logs/_tests_cv_classification_hydra
+LOGFILE=${LOGDIR}/checkpoints/_metrics.json
+
+PYTHONPATH=./examples:./catalyst:${PYTHONPATH} \
```
you need to remove `./catalyst/` to resolve typing issue, Catalyst also has `catalyst/typing.py`
```suggestion
PYTHONPATH=./examples:.:${PYTHONPATH} \
```

otherman16 replied:
Thanks. It works now! =)

Scitator commented on catalyst/experiments/hydra_config.py:
```diff
@@ -0,0 +1,552 @@
+from typing import Any, Dict, List, Mapping
```
dear @otherman16 ,
could we move all duplicated from `ConfigExperiment` to `catalyst/experiments/functional.py`?
or, is it possible to do something like `HydraConfigExperiment(ConfigExperiment)`?

otherman16 replied:
I can only move functions `process_callbacks()` and `add_default_callbacks()` to `catalyst/experiments/functional.py`.
The rest of the code are only almost similar.

Scitator commented on catalyst/runners/hydra_supervised.py:
```diff
@@ -0,0 +1,195 @@
+logger = logging.getLogger(__name__)
+
+
+class HydraSupervisedRunner(Runner):
```
do we really need it?
from pipeline perspective (model & data flow)  - there are no changes, is't it?
from Config API - we could redefine used experiment with
```
experiment_params:
  ...
```
like https://github.com/catalyst-team/catalyst/blob/master/examples/configs/config-description-eng.yml#L26, but `experiment_params` :joy:

Scitator replied:
another option - `HydraSupervisedRunner(SupervisedRunner)`

otherman16 replied:
Generally `HydraSupervisedRunner` is very different from `SupervisedRunner`.
1. __init__ parameters: `model_keys` except `input_key/output_key/input_target_key`
2. Multi-model support: there are separate `input_key/output_key/input_target_key` for each `model`
3. _process_input/_process_output/_handle_device/forward methods are different
But think I can inherit `HydraSupervisedRunner` from `SupervisedRunner`.
I wanted to minimize dependency between `HydraSupervisedRunner` and `SupervisedRunner`. But if you say so I rewrite.

Scitator replied:
let's rename it to `MultiSupervisedRunner`? 

I also think that we should use `AutoCallbackExperiment` as the default.
`experiment_fn: Callable` is good for notebook experiments, but with Config API you could easily rewrite the default Experiment with 
```
experiment_params:
    ...
```
or just write it into `__init__.py`

Scitator reviewed (commented):
thrilled by this PR, amazing work!

omry commented on catalyst/utils/hydra_config.py:
```diff
@@ -0,0 +1,52 @@
+
+    cfg.setdefault("vals", DictConfig({}))
+
+    cfg.setdefault("args", DictConfig({}))
+    cfg.args.setdefault("expdir", ".")
+    cfg.args.setdefault("resume", None)
+    cfg.args.setdefault("autoresume", None)
+    cfg.args.setdefault("seed", 42)
+    cfg.args.setdefault("distributed", os.getenv("USE_DDP", "0") == "1")
+    cfg.args.setdefault("apex", os.getenv("USE_APEX", "0") == "1")
+    cfg.args.setdefault("amp", os.getenv("USE_AMP", "0") == "1")
```
Thanks for working on this!
While I understand you have backward compatibility requirements that might be in conflict with the suggestions here, I wanted to provide them anyway (at least as food for thought) :
I'd recommend utilizing config composition to get this instead of manually editing the config.
for example, the defaults list of the primary config can look like:
```yaml
defaults:
  - catalyst/config  # mandatory catalyst config (provided by catalyst)
  - foo: bar # user config groups
```
catalyst/config can contain everything needed by catalyst (what you are adding above).

A few extra points:
1. I am guessing this is designed to be backward compatible with the standard command line.
If possible, break away from that and utilize hierarchy to give catalyst a space of it's own where it wouldn't conflict with other frameworks/user config.
for example:
*catalst/config.yaml*:
```yaml
catalyst:
  resume: null
  seed: 42
  ...
```

2. Consider using Structured Configs to get config type safety, the above yaml file can be replaced with something like:
```python
@dataclass
class CatalystConf:
  resume: Optional[bool] = None
  seed: int = 42

ConfigStore.instance().store(group="catalst", name="config", node=CatalystConf)
```
(See the Structured Configs tutorial for more info).

otherman16 replied:
Thank you. Thinking about it.

Scitator commented on catalyst/experiments/hydra_config.py:
```diff
@@ -0,0 +1,552 @@
+            model_params = []
+            for model_key_el in model_keys:
+                model_params_el = process_model_params(
+                    models[model_key_el],
```
could we move that to functional?

otherman16 replied:
Yes we can, but it will not be elegant.

Scitator commented on catalyst/experiments/hydra_config.py:
```diff
@@ -0,0 +1,552 @@
+            # update optimizer params
+            for key, value in params.items():
+                for optimizer_param_group in optimizer.param_groups:
+                    optimizer_param_group[key] = value
```
to functional also?

otherman16 replied:
Yes we can.

Scitator commented on catalyst/experiments/hydra_config.py:
```diff
@@ -0,0 +1,552 @@
+        if "optimizers" not in self._config.stages[stage]:
+            return {}
+        optimizers_params: DictConfig = self._config.stages[stage].optimizers
+        optimizer = {
+            name: self._get_optimizer(stage, model, name, optimizer_params)
+            for name, optimizer_params in optimizers_params.items()
+        }
```
is there any support for non-key-value optimizer?

otherman16 replied:
No. All are in key-value style (models, optimizers, schedulers, datasets, loaders, transforms, callbacks).
We do not need "key_val" flag in this case. Think this method is much more simple and stable because of the unification.

Scitator commented on catalyst/experiments/hydra_config.py:
```diff
@@ -0,0 +1,552 @@
+        if "criterions" not in self._config.stages[stage]:
+            return {}
+        criterions_params: DictConfig = self._config.stages[stage].criterions
+        criterion: Dict[str, Criterion] = {
+            name: self._get_criterion(criterion_params)
+            for name, criterion_params in criterions_params.items()
+        }
```
is there any support for non-key-value criterion?

otherman16 replied:
No. See previous answer.

Scitator commented on catalyst/experiments/hydra_config.py:
```diff
@@ -0,0 +1,552 @@
+        if "schedulers" not in self._config.stages[stage]:
+            return {}
+        schedulers_params: DictConfig = self._config.stages[stage].schedulers
+        scheduler: Dict[str, Scheduler] = {
+            key: self._get_scheduler(
+                optimizers=optimizer, params=scheduler_params
+            )
+            for key, scheduler_params in schedulers_params.items()
+        }
```
is there any support for non-key-value scheduler?

otherman16 replied:
No. See previous answer.

Scitator commented on catalyst/runners/__init__.py:
```diff
@@ -3,4 +3,11 @@
+
+    __all__ = ["Runner", "SupervisedRunner", "HydraSupervisedRunner"]
+else:
+    __all__ = ["Runner", "SupervisedRunner"]
```
In the case of renaming we also could simply this to:
```
__all__ = ["Runner", "SupervisedRunner", "MultiSupervisedRunner"]
```

otherman16 replied:
Ok. It is possible.

Scitator commented on catalyst/utils/hydra_scripts.py:
```diff
@@ -0,0 +1,70 @@
+
+def distributed_cmd_run(
+    worker_fn: Callable, distributed: bool = True, cfg: DictConfig = None
+) -> None:
```
do we really need it? current ddp cmd:
```
def distributed_cmd_run(
    worker_fn: Callable, distributed: bool = True, *args, **kwargs
) -> None:
```

otherman16 replied:
I'll try.

Scitator commented on catalyst/utils/hydra_sys.py:
```diff
@@ -0,0 +1,64 @@
+logger = logging.getLogger(__name__)
+
+
+def dump_environment(cfg: DictConfig, logdir: str) -> None:
```
how we could merge it with the current `dump_environment` implementation?
is it possible to just add extra hydra mixins with `IS_HYDRA_AVAILABLE` flag?

otherman16 replied:
I'll try.

otherman16 committed:
Update check_dl_cv.sh
Hydra cv classification test was fixed

otherman16 committed:
Update config.py, functional.py, and hydra_config.py
All dublicated code was replaced to catalyst/experiments/functional.py

otherman16 committed:
Update hydra_supervised.py
Hydra cv classification test was fixed

otherman16 committed:
Removed catalyst/utils/hydra_scripts.py, catalyst/utils/hydra_sys.py and changed catalyst/utils/sys.py:
Functions 'catalyst/utils/scripts.py:distributed_cmd_run()' and 'catalyst/utils/sys.py:dump_environment()' are used for both HydraConfigExperiment and ConfigExperiment

otherman16 committed:
Renamed catalyst/runners/hydra_supervised.py to catalyst/runners/multi_supervised.py
MultiSupervisedRunner was added for multiple model support

otherman16 committed:
Update config_infer.yaml and config_train.yaml due to MultiSupervisedRunner

otherman16 committed:
Updated Dockerfile-dev and Dockerfile-dev-fp16 with installing requirements/requirements-hydra.txt

otherman16 committed:
Update check_dl_cv.sh and run.py for hydra-tests fix

otherman16 committed:
Update config.py and hydra_config.py

Scitator commented on catalyst/utils/hydra_config.py:
```diff
@@ -0,0 +1,52 @@
+from omegaconf import DictConfig, OmegaConf
+
+
+def prepare_config(cfg: DictConfig) -> DictConfig:
```
maybe, `prepare_hydra_config`? and resolve naming?

otherman16 replied:
ok

otherman16 committed:
catalyst/utils/hydra_config.py: prepare_config() -> prepare_hydra_config()

Scitator committed:
Merge branch 'master' into feature/hydra
